src/util/CompositeAsyncHandler.ts

```diff
@@ -0,0 +1,86 @@
+import { AsyncHandler } from './AsyncHandler';
+import { UnsupportedHttpError } from './errors/UnsupportedHttpError';
+
+/**
+ * Handler that combines several other handlers,
+ * thereby allowing other classes that depend on a single handler to still use multiple.
+ */
+export class CompositeAsyncHandler<TIn, TOut> implements AsyncHandler<TIn, TOut> {
+  private readonly handlers: AsyncHandler<TIn, TOut>[];
+
+  /**
+   * Creates a new CompositeAsyncHandler that stores the given handlers.
+   * @param handlers - Handlers over which it will run.
+   */
+  public constructor (handlers: AsyncHandler<TIn, TOut>[]) {
+    this.handlers = handlers;
+  }
+
+  /**
+   * Checks if any of the stored handlers can handle the given input.
+   * @param input - The data that would need to be handled.
+   *
+   * @returns A promise resolving if at least 1 handler supports to input, or rejecting if none do.
+   */
+  public async canHandle (input: TIn): Promise<void> {
+    await this.findHandler(input);
+  }
+
+  /**
+   * Finds a handler that supports the given input and then lets it handle the given data.
+   * @param input - The data that needs to be handled.
+   *
+   * @returns A promise corresponding to the handle call of a handler that supports the input.
+   * It rejects if no handlers support the given data.
+   */
+  public async handle (input: TIn): Promise<TOut> {
+    let handler: AsyncHandler<TIn, TOut>;
+
+    try {
+      handler = await this.findHandler(input);
+    } catch (error) {
+      throw new Error('All composed handlers failed. This might be the consequence of calling handle before canHandle.');
+    }
+
+    return handler.handle(input);
+  }
+
+  /**
+   * Identical to {@link AsyncHandler.handleSafe} but optimized for composite by only needing 1 canHandle call on members.
+   * @param input - The input data.
+   *
+   * @returns A promise corresponding to the handle call of a handler that supports the input.
+   * It rejects if no handlers support the given data.
+   */
+  public async handleSafe (input: TIn): Promise<TOut> {
+    const handler = await this.findHandler(input);
+
+    return handler.handle(input);
+  }
+
+  /**
+   * Finds a handler that can handle the given input data.
+   * Otherwise an error gets thrown.
+   *
+   * @param input - The input data.
+   *
+   * @returns A promise resolving to a handler that supports the data or otherwise rejecting.
+   */
+  private async findHandler (input: TIn): Promise<AsyncHandler<TIn, TOut>> {
+    const errors: Error[] = [];
+
+    for (const handler of this.handlers) {
+      try {
+        await handler.canHandle(input);
+
+        return handler;
+      } catch (error) {
+        errors.push(error);
+      }
+    }
+
+    const joined = errors.map((error: Error): string => error.message).join(', ');
+
+    throw new UnsupportedHttpError(`No composed handler supports the given input: [${joined}].`);
```

Comment:
I find it weird that this is an HTTP error.

Reply:
So @rubensworks, after offline discusion with @joachimvh, I think we really see the drawback here of having `HTTP` in the name. The `CompositeAsyncHandler` is so generic is doesn't have anything to do with Solid, let alone with HTTP.

I think the fact that there is an HTTP status code hidden in the error should not leak to components; also so we can change that mechanism later. From the perspective of this handler, it's just not supported; it doesn't care about the mechanism we use to eventually turn this into an HTTP error.

Reply:
I partially agree/disagree.

> The CompositeAsyncHandler is so generic is doesn't have anything to do with Solid, let alone with HTTP.

I agree with this.

However, since this error _does_ have an implicit link to HTTP (via its encoded status code), the class name should still represent this IMO. If the class name should become `UnsupportedError` (which I would be fine with), then the status code should not be encoded in it IMO. Otherwise this could lead to unforeseen consequences.

If we really don't want `CompositeAsyncHandler` to have anything to do with Solid, then we should be able to (theoretically) move it in a separate package. And since that package won't have any links to HTTP, it wouldn't be able to make use of this specific error class that has an HTTP status code encoded in it.

This also makes me realize that we actually do need a special system to translate regular errors into HTTP errors, because we could make use of external libraries that throw their own kinds of errors.

-----

One cleaner alternative may be to accept an error handler/constructor in `CompositeAsyncHandler`'s constructor. When we then instantiate `CompositeAsyncHandler` in an HTTP context, we could pass `UnsupportedHttpError`'s constructor to it.

Reply:
> then the status code should not be encoded in it IMO

Perhaps agree; but I would consider this an internal implementation detail at the moment. It just _happens_ to be the kind of error that translates easily into HTTP, but no one needs to know that at this point.

> This also makes me realize that we actually do need a special system to translate regular errors into HTTP errors

Indeed; for now, this system is simple (it's hidden inside of the error), but we might want to make that a dedicated component in the long run. Which is precisely why I want to hide the `HTTP` part of the error externally. For now, that dedicated component can be simple then: if has status code, then use that one; if not, 500.

> One cleaner alternative may be to accept an error handler/constructor

That seems overkill.

Reply:
> but I would consider this an internal implementation detail at the moment. It just happens to be the kind of error that translates easily into HTTP, but no one needs to know that at this point.

I disagree on this point. But I'm not sure we'll be able to convince each other.
Since this is too much of a detail to spend too much time on, let's agree to disagree.

But ok, my concerns are on the record here.
You can make the final decision on the naming @RubenVerborgh :-)

Reply:
Thanks; I'll defer to @joachimvh then 🙂

Reply:
I don't want to do another rebase so I'm going to leave it as is ;) . Maybe we need to open an issue for this discussion since at this point it's clear that this will probably change in the future.